src/uscheduled.8.in
.TH uscheduled 8 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uscheduled - a job scheduler
.LINE
.section SYNOPSIS
\fBuscheduled\fR [-d DIR]
.section DESCRIPTION
\fBuscheduled\fR reads information about a number of jobs 
from \fIDIR\fR, or, if this is not given, from \fI~/.uschedule/\fR.
.paragraph
It then sleeps until one or more of the jobs have to be run, at which
point it will try to run and either reschedule or delete them.
.paragraph
\fBuscheduled\fR runs under the rights of the
user for whom the programs are started. It changes to \fIDIR\fR on
startup and runs in an own session. It doesn't detach from the
terminal or move itself into the background.
.paragraph
All time calculation is done in UTC.
.section "JOB EXECUTION"
If \fBuscheduled\fR needs to start a job the following will be 
done:
.list
.entry "uscheduled forks a new process."
.entry "the new process redirects /dev/null"
to file descriptors one and two, if desired.
.entry "the new process changes to $HOME."
.entry "the new process executes the program."
.entry "uscheduled reschedules the job.
.endlist
The users program will inherit the environment of the daemon, 
unless otherwise indicated during job creation time.
The standard input will be redirected from /dev/null. By default
the file descriptors one and two (standard and error output) are
not redirected, meaning that the output will go into the same
log files as the output of the daemon, if any.
.section "SIGNAL HANDLING"
\fBuscheduled\fR exits gracefully on SIGTERM. It reloads the job
specifications on SIGHUP.
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedule.html uschedule(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uscheduled.c
#include "scheduled.h"
#include "stralloc.h"
#include "sig.h"
#include "iopause.h"
#include "byte.h"
#include "str.h"
#include "uogetopt.h"
#include "open.h"
#include "pathexec.h"
#include "bailout.h"
#include "error.h"
#include "wait.h"
#include "fmt.h"
#include "coe.h"
#include "env.h"
#include "uolock.h"
#include "ndelay.h"
#include "wrap_stat.h"
#include <unistd.h>

/* he who decided that rename should be placed in stdio.h should be ashamed */
extern int rename(const char *oldpath, const char *newpath);

static int got_sigterm;
static int got_sigchld;
static int got_sighup;
static void sigterm_handler(int st) {got_sigterm=st; }
static void sighup_handler(int st) {got_sighup=st; }
static void sigchld_handler(int st) {got_sigchld=st; }

static const char *home;
static stralloc absprefix;
static stralloc jobs;

static void 
delete(const char *s, struct jobinfo *j)
{
	static stralloc id;

	if (!stralloc_copyb(&id,j->id,j->idlen)) oom();
	if (!stralloc_0(&id)) oom();
	if (-1==unlink(s) && errno!=error_noent) 
		warning(errno,"failed to unlink stamp of job",id.s,0,0);
	else
		warning(0,"deleted job ",id.s,0,0);
}

static int 
reschedule(const char *s,struct jobinfo *j)
{
	static stralloc sa;
	struct taia now, next;

	taia_now(&now);

	if (1==j->repeats) {
		delete(s,j);
		return 0;
	}
	if (j->repeats)
		j->repeats--;

	j->lastrun=now;

	if (!find_next(j,&now,&next)) {
		delete(s,j);
		return 0;
	}

	make_name(&sa,j);

	if (0==rename(s,sa.s)) {
		warning(0,"rescheduled ", s," to ", sa.s);
		return 1;
	}
	warning(errno,"failed to rename ", s , " to ", sa.s);
	return 0;
}

static void 
execute(const char *s, struct jobinfo *ji)
{
	int pid;
	pid=fork();
	if (-1==pid) {
		warning(errno,"failed to fork, job delayed",0,0,0);
		sleep(1);
		return;
	}
	if (0==pid) {
		/* child */
		static stralloc c;
		char *av[2];
		int fd;
		if (!stralloc_copy(&c,&absprefix)) oom();
		if (!stralloc_cats(&c,"/")) oom();
		if (!stralloc_cats(&c,IDDIR)) oom();
		if (!stralloc_cats(&c,"/")) oom();
		if (!stralloc_catb(&c,ji->id,ji->idlen)) oom();
		if (!stralloc_0(&c)) oom();
		if (!pathexec_env("SCHEDULED_COMMAND_FILE",c.s)) oom();
		if (ji->null1) {
			close(1);
			if (-1==open_write("/dev/null"))
				xbailout(111,errno,"failed to open /dev/null",0,0,0);
		}
		if (ji->null2) {
			close(2);
			if (-1==open_write("/dev/null"))
				xbailout(111,errno,"failed to open /dev/null",0,0,0);
		}
		av[0]=c.s;
		av[1]=0;
		if (-1==chdir(home))
			xbailout(111,errno,"failed to chdir to ",home,0,0);
		fd=open_read(c.s);
		if (-1==fd)
			xbailout(111,errno,"failed to open_read ",c.s,0,0);
		coe(fd);
		/* to allow editing of command file */
		if (-1==uolock_share(fd))
			xbailout(111,errno,"failed to lock ",c.s,0,0);
		pathexec(av);
		if (errno==error_noent) {
			struct wrap_stat st;
			int e=errno;
			/* don't ever trust kernels errno after execv ... */
			if (-1==wrap_stat(av[0],&st) && errno==error_noent) {
				/* there's an entry in the schedule directory, but no
				 * matching entry in the command directory. This can
				 * happen if someone used rm or if schedulerm lost the
				 * race.
				 */
				warning(errno,"delete entry ", s,0,0);
				delete(s,ji);
				_exit(1);
			}
			errno=e;
		}
		warning(errno,"failed to execute ", av[0],0,0);
		_exit(1);
	} else {
		/* parent */
		char nb[FMT_ULONG];
		nb[fmt_ulong(nb,pid)]=0;
		warning(0,"started pid ",nb, " for job ",s);
		reschedule(s,ji);
	}
}

static const char *o_dir;
static uogetopt2 myopts[]={
{'d',"dir",uogo_string,0,&o_dir, 0 ,
"Jobdirectory.",
"Jobs will be read from this directory.","DIR"},
{0,0,0,0,0,0,0,0,0}
};
static uogetopt_env myoptenv={
  "uscheduled",PACKAGE,VERSION,
  "uscheduled [-d DIR]",
  "  This program schedules the content of DIR or ~/.uschedule.",
  "long",
  "  Report bugs to uschedule@lists.ohse.de",
  1,1,
  0,0,uogetopt_out,myopts
};

static iopause_fd iop[1];
static void fifo_open(void) 
{
	iop[0].fd=open_read("fifo");
	if (-1==iop[0].fd)
		xbailout(111,errno,"failed to open fifo",0,0,0);
	ndelay_on(iop[0].fd);
	iop[0].events=IOPAUSE_READ;
}
static void fifo_read(void)
{
	while (1) {
		int r;
		char c;
		r=read(iop[0].fd,&c,1);
		if (-1==r)
			break;
		if (0==r)
			break;
	}
	close(iop[0].fd);
	fifo_open();
}
static void reap(void)
{
	do {
		int st;
		int pid;
		char nb1[FMT_ULONG];
		char nb2[FMT_ULONG];
		pid=wait_nohang(&st);
		if (-1==pid || 0==pid)
			break;
		nb1[fmt_ulong(nb1,pid)]=0;
		nb2[fmt_ulong(nb2,st)]=0;
		warning(0,"processs ",nb1, " terminated with code ",nb2);
	} while (1);
}

static void 
calc_next_run(struct taia *now, struct taia *then)
{
  unsigned int i;
  taia_uint(then,1800); /* sleep no more than 30 minutes */
  taia_add(then,now,then);
  for (i=0;jobs.s[i];i+=str_len(jobs.s+i)+1) {
    struct jobinfo ji;
    struct taia nextrun;
    /* load_jobs did the checks on the job */
    parse_job(jobs.s+i,&ji);
    if (find_next(&ji,now, &nextrun))
      if (taia_less(&nextrun,then))
	*then=nextrun;
    /* care about timed-out jobs below */
  }
  if (taia_less(then,now))
	  *then=*now;
}

static void 
do_executes(struct taia *now)
{
  unsigned int i;
  for (i=0;jobs.s[i];i+=str_len(jobs.s+i)+1) {
    struct taia nextrun;
    struct jobinfo ji;
    parse_job(jobs.s+i,&ji); /* load_jobs did the checks on the job */
    if (!find_next(&ji,now,&nextrun)) {
      warning(0,"no next run, deleting ",jobs.s+i,0,0);
      delete(jobs.s+i,&ji);
      continue;
    }
    if (taia_less(now,&nextrun))
      continue;
    execute(jobs.s+i,&ji);
  }
}

int 
main(int argc, char **argv)
{
	bailout_progname(argv[0]);
	flag_bailout_fatal_begin=3;
	flag_bailout_log_pid=1;
	uogo_posixmode=1;
	myoptenv.program=flag_bailout_log_name;
	uogetopt_parse(&myoptenv,&argc,argv);
	change_dir(&absprefix, o_dir,0);
	home=env_get("HOME"); /* change_dir checks for existance */

	warning(0,"starting",0,0,0);

	sig_catch(sig_term,sigterm_handler);
	sig_catch(sig_hangup,sighup_handler);
	sig_catch(sig_child,sigchld_handler);
	sig_ignore(sig_pipe);
	setsid();
	close(0);
	if (-1==open_read("/dev/null"))
		xbailout(111,errno,"failed to open /dev/null",0,0,0);
	fifo_open();

	while (1) {
		struct taia now,then;
		if (got_sigterm) break;
		if (got_sigchld) {
			reap();
			got_sigchld=0;
		}
		if (got_sighup) {
			warning(0,"reloading",0,0,0);
			got_sighup=0;
		}
		load_jobs(".",&jobs); 

		taia_now(&now);

		calc_next_run(&now,&then);

		iopause(iop,1,&then,&now);
		if (iop[0].revents)
			fifo_read();

		taia_now(&now);

		if (taia_less(&now,&then))
			continue; /* signal */

		do_executes(&now);
	}
	warning(0,"exiting",0,0,0);
	_exit(0);
}
